micrometer-core/src/main/java/io/micrometer/core/instrument/binder/CaffeineCacheMetrics.java

```diff
@@ -0,0 +1,90 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.micrometer.core.instrument.binder;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tag;
+import io.micrometer.core.instrument.Tags;
+
+import java.util.List;
+
+/* Collect metrics from Caffeine's com.github.benmanes.caffeine.cache.Cache.
+  * <p>
+  * <pre>{@code
+  *
+  * // Note that `recordStats()` is required to gather non-zero statistics
+  * Cache<String, String> cache = Caffeine.newBuilder().recordStats().build();
+  * CacheMetricsCollector cacheMetrics = new CacheMetricsCollector().register();
+  * cacheMetrics.addCache("mycache", cache);
+  *
+  * }</pre>
+  *
+  * Exposed metrics are labeled with the provided cache name.
+  *
+  * With the example above, sample metric names would be:
+  * <pre>
+  *     caffeine_cache_hit_total{cache="mycache"} 10.0
+  *     caffeine_cache_miss_total{cache="mycache"} 3.0
+  *     caffeine_cache_requests_total{cache="mycache"} 13.0
+  *     caffeine_cache_eviction_total{cache="mycache"} 1.0
+  *     caffeine_cache_estimated_size{cache="mycache"} 5.0
+  * </pre>
+  *
+  * Additionally if the cache includes a loader, the following metrics would be provided:
+  * <pre>
+  *     caffeine_cache_load_failure_total{cache="mycache"} 2.0
+  *     caffeine_cache_loads_total{cache="mycache"} 7.0
+  *     caffeine_cache_load_duration_seconds_count{cache="mycache"} 7.0
+  *     caffeine_cache_load_duration_seconds_sum{cache="mycache"} 0.0034
+  * </pre>
+  *
+  * @author Clint Checketts
+ */
+public class CaffeineCacheMetrics implements MeterBinder {
+  private final String name;
+  private final Cache<?, ?> cache;
+
+  /**
+   * @param name  The named of the cache, exposed as the 'cache' dimension
+   * @param cache The cache to be instrumented. Be certain to enable <cade></cade>
+   */
+  public CaffeineCacheMetrics(String name, Cache<?, ?> cache) {
+    this.name = name;
+    this.cache = cache;
+  }
+
+  @Override
+  public void bindTo(MeterRegistry registry) {
+    List<Tag> tags = Tags.zip("cache", name);
+    registry.gauge("caffeine_cache_estimated_size", tags, cache, Cache::estimatedSize);
+
+    registry.counter("caffeine_cache_requests", Tags.zip("cache", name, "result", "miss"), cache, c -> c.stats().missCount());
+    registry.counter("caffeine_cache_requests", Tags.zip("cache", name, "result", "hit"), cache, c -> c.stats().hitCount());
+    registry.counter("caffeine_cache_requests_total", tags, cache, c -> c.stats().requestCount());
+    registry.counter("caffeine_cache_evictions_total", tags, cache, c -> c.stats().evictionCount());
+    registry.gauge("caffeine_cache_eviction_weight", tags, cache, c -> c.stats().evictionWeight());
+
+    if (cache instanceof LoadingCache) {
```

Comment:
@ben-manes  Are you saying this line should be `cache instanceof LoadingCache || cache instanceof AsyncLoadingCache`  ?